File: android/app/src/main/kotlin/ec/cacaotrace/datos/repositorios/RepositorioProduccion.kt
package ec.cacaotrace.datos.repositorios

import androidx.room.withTransaction
import ec.cacaotrace.datos.bd.BaseDatos
import ec.cacaotrace.datos.bd.Comunes
import ec.cacaotrace.datos.bd.entidades.AtemperadoEntidad
import ec.cacaotrace.datos.bd.entidades.DescascarilladoEntidad
import ec.cacaotrace.datos.bd.entidades.EmpaqueEntidad
import ec.cacaotrace.datos.bd.entidades.LoteProduccionEntidad
import ec.cacaotrace.datos.bd.entidades.LoteProduccionOrigenEntidad
import ec.cacaotrace.datos.bd.entidades.MovimientoInventarioEntidad
import ec.cacaotrace.datos.bd.entidades.RefinadoEntidad
import ec.cacaotrace.datos.bd.entidades.TostadoEntidad
import ec.cacaotrace.datos.sync.ServicioSincronizacion
import ec.cacaotrace.nucleo.calculo.CalculadoraReceta
import ec.cacaotrace.nucleo.calculo.Receta
import ec.cacaotrace.nucleo.modelo.Codigos
import ec.cacaotrace.nucleo.modelo.EstadoSync
import ec.cacaotrace.nucleo.modelo.MetodoAtemperado
import ec.cacaotrace.nucleo.reglas.Alerta
import ec.cacaotrace.nucleo.reglas.ContextoAtemperado
import ec.cacaotrace.nucleo.reglas.ContextoTostado
import ec.cacaotrace.nucleo.reglas.MotorReglas
import java.time.Instant
import java.time.ZoneId
import kotlinx.coroutines.flow.Flow
import kotlinx.serialization.json.Json

/** Repositorio del lote de producción: del grano seco a la barra empacada. */
class RepositorioProduccion(
    private val bd: BaseDatos,
    private val sync: ServicioSincronizacion,
    private val config: RepositorioConfiguracion,
    private val alertas: RepositorioAlertas,
) {
    private val json = Json { encodeDefaults = true }
    private val dao get() = bd.produccion()

    private suspend fun motor() = MotorReglas(config.umbrales())

    fun observar(): Flow<List<LoteProduccionEntidad>> = dao.observarTodas()

    suspend fun porId(id: String): LoteProduccionEntidad? = dao.porId(id)

    /**
     * Crea una tanda a partir de uno o varios lotes de grano (RF-LOT-05).
     *
     * Descuenta el grano del inventario, porque salir de almacén hacia
     * producción es un movimiento real (RF-ALM-03). Y se niega a usar un lote
     * bloqueado por laboratorio: dejar pasar cadmio a producción convertiría
     * un lote perdido en una tanda entera perdida (RN-15).
     */
    suspend fun crear(
        kgPorLote: Map<String, Double>,
        porcentajeCacao: Double = 90.0,
    ): LoteProduccionEntidad {
        require(kgPorLote.isNotEmpty()) { "Hay que indicar al menos un lote de grano" }

        for (loteId in kgPorLote.keys) {
            alertas.bloqueoActivo(loteId)?.let { throw LoteBloqueado(loteId, it.quePaso) }
        }

        val ahora = Instant.now()
        val anio = ahora.atZone(ZoneId.systemDefault()).year
        val codigo = Codigos.siguiente(Codigos.PREFIJO_PRODUCCION, anio, dao.codigosUsados())

        val tanda = LoteProduccionEntidad(
            codigo = codigo,
            fecha = ahora,
            porcentajeCacao = porcentajeCacao,
        )

        bd.withTransaction {
            dao.insertar(tanda)
            for ((loteId, kg) in kgPorLote) {
                dao.insertarOrigen(
                    LoteProduccionOrigenEntidad(
                        loteProduccionId = tanda.id,
                        loteId = loteId,
                        kgUsados = kg,
                    ),
                )
                bd.inventario().insertarMovimiento(
                    MovimientoInventarioEntidad(
                        item = "grano_seco",
                        tipo = "salida",
                        cantidad = kg,
                        fecha = ahora,
                        referencia = codigo,
                    ),
                )
            }
        }
        sync.encolar("lotes_produccion", tanda.id, "crear", """{"codigo":"$codigo"}""")
        return tanda
    }

    suspend fun completa(id: String): ProduccionCompleta? {
        val p = porId(id) ?: return null
        return ProduccionCompleta(
            produccion = p,
            origenes = dao.origenes(id),
            tostado = dao.tostado(id),
            descascarillado = dao.descascarillado(id),
            refinado = dao.refinado(id),
            atemperados = dao.atemperados(id),
            empaque = dao.empaque(id),
        )
    }

    // ------------------------------------------------------------ tostado

    /** Registra el tostado y evalúa la merma (RN-12). */
    suspend fun guardarTostado(
        loteProduccionId: String,
        kgEntrada: Double,
        equipoId: String? = null,
        kgSalida: Double? = null,
        tempC: Double? = null,
        minutos: Int? = null,
        perfil: List<Map<String, Double>> = emptyList(),
        gradoIa: String = "",
        confianzaIa: Double? = null,
        gradoUsuario: String = "",
        modeloVersion: String = "",
        fotoId: String? = null,
        recetaNombre: String = "",
    ): List<Alerta> {
        val existente = dao.tostado(loteProduccionId)
        val tostado = (existente ?: TostadoEntidad(
            loteProduccionId = loteProduccionId,
            fecha = Instant.now(),
        )).copy(
            equipoId = equipoId,
            kgEntrada = kgEntrada,
            kgSalida = kgSalida,
            tempC = tempC,
            minutos = minutos,
            perfilJson = json.encodeToString(perfil),
            gradoIa = gradoIa,
            confianzaIa = confianzaIa,
            gradoUsuario = gradoUsuario,
            modeloVersion = modeloVersion,
            fotoId = fotoId,
            recetaNombre = recetaNombre,
            comunes = (existente?.comunes ?: Comunes()).tocada(),
        )
        dao.guardarTostado(tostado)
        sync.encolar("tostados", tostado.id, "actualizar")

        if (kgSalida == null) return emptyList()
        return alertas.registrar(
            motor().evaluarTostado(ContextoTostado(kgEntrada, kgSalida)),
            loteProduccionId = loteProduccionId,
        )
    }

    // ------------------------------------------------------- descascarillado

    suspend fun guardarDescascarillado(
        loteProduccionId: String,
        kgNibs: Double,
        kgCascarilla: Double,
        notas: String = "",
    ): List<Alerta> {
        val existente = dao.descascarillado(loteProduccionId)
        val d = (existente ?: DescascarilladoEntidad(
            loteProduccionId = loteProduccionId,
            fecha = Instant.now(),
        )).copy(
            kgNibs = kgNibs,
            kgCascarilla = kgCascarilla,
            notas = notas,
            comunes = (existente?.comunes ?: Comunes()).tocada(),
        )
        dao.guardarDescascarillado(d)
        sync.encolar("descascarillados", d.id, "actualizar")

        bd.inventario().insertarMovimiento(
            MovimientoInventarioEntidad(
                item = "nibs",
                tipo = "entrada",
                cantidad = kgNibs,
                fecha = Instant.now(),
                referencia = loteProduccionId,
            ),
        )

        val tostado = dao.tostado(loteProduccionId)
        val kgSalida = tostado?.kgSalida ?: return emptyList()
        return alertas.registrar(
            motor().evaluarTostado(
                ContextoTostado(tostado.kgEntrada, kgSalida, kgNibs, kgCascarilla),
            ),
            loteProduccionId = loteProduccionId,
        )
    }

    // ------------------------------------------------------------ refinado

    /** Calcula la receta para los nibs disponibles (RF-REF-01). */
    fun calcularReceta(
        kgNibs: Double,
        porcentajeCacao: Double = 90.0,
        mantecaExtraPct: Double = 0.0,
        usarLecitina: Boolean = true,
    ): Receta = CalculadoraReceta().calcular(
        kgNibs = kgNibs,
        porcentajeCacao = porcentajeCacao,
        mantecaExtraPct = mantecaExtraPct,
        usarLecitina = usarLecitina,
    )

    suspend fun guardarRefinado(
        loteProduccionId: String,
        inicio: Instant,
        nibsKg: Double,
        azucarKg: Double,
        fin: Instant? = null,
        horas: Double? = null,
        mantecaKg: Double = 0.0,
        lecitinaKg: Double = 0.0,
        momentoAzucar: Instant? = null,
        tempC: Double? = null,
        sensorial: Map<String, Int> = emptyMap(),
        notas: String = "",
    ): String {
        val existente = dao.refinado(loteProduccionId)
        val r = (existente ?: RefinadoEntidad(
            loteProduccionId = loteProduccionId,
            inicio = inicio,
        )).copy(
            inicio = inicio,
            fin = fin,
            horas = horas,
            nibsKg = nibsKg,
            azucarKg = azucarKg,
            mantecaKg = mantecaKg,
            lecitinaKg = lecitinaKg,
            momentoAzucar = momentoAzucar,
            tempC = tempC,
            sensorialJson = json.encodeToString(sensorial),
            notas = notas,
            comunes = (existente?.comunes ?: Comunes()).tocada(),
        )
        dao.guardarRefinado(r)
        sync.encolar("refinados", r.id, "actualizar")

        // El azúcar, la manteca y la lecitina salen del inventario al usarse.
        mapOf("azucar" to azucarKg, "manteca" to mantecaKg, "lecitina" to lecitinaKg)
            .filterValues { it > 0 }
            .forEach { (item, cantidad) ->
                bd.inventario().insertarMovimiento(
                    MovimientoInventarioEntidad(
                        item = item,
                        tipo = "salida",
                        cantidad = cantidad,
                        fecha = Instant.now(),
                        referencia = loteProduccionId,
                    ),
                )
            }
        return r.id
    }

    // ------------------------------------------------------------ atemperado

    /** Registra el atemperado y evalúa RN-13 y RN-14. */
    suspend fun guardarAtemperado(
        loteProduccionId: String,
        metodo: MetodoAtemperado = MetodoAtemperado.SIEMBRA,
        temperaturas: Map<String, Double> = emptyMap(),
        tempCuarto: Double? = null,
        hrCuarto: Double? = null,
        pruebaPapel: Boolean? = null,
        resultadoIa: String = "",
        confianzaIa: Double? = null,
        resultadoUsuario: String = "",
        modeloVersion: String = "",
        fotoId: String? = null,
        diasInspeccion: Int? = null,
    ): List<Alerta> {
        val a = AtemperadoEntidad(
            loteProduccionId = loteProduccionId,
            fecha = Instant.now(),
            metodo = metodo,
            tempsJson = json.encodeToString(temperaturas),
            tempCuarto = tempCuarto,
            hrCuarto = hrCuarto,
            pruebaPapel = pruebaPapel,
            resultadoIa = resultadoIa,
            confianzaIa = confianzaIa,
            resultadoUsuario = resultadoUsuario,
            modeloVersion = modeloVersion,
            fotoId = fotoId,
            diasInspeccion = diasInspeccion,
        )
        dao.insertarAtemperado(a)
        sync.encolar("atemperados", a.id, "crear")

        return alertas.registrar(
            motor().evaluarAtemperado(
                ContextoAtemperado(tempCuarto, hrCuarto, temperaturas["trabajo"]),
            ),
            loteProduccionId = loteProduccionId,
        )
    }

    /** ¿Se puede moldear ahora mismo en estas condiciones? (RF-ATE-02) */
    suspend fun advertenciaDelCuarto(tempCuarto: Double?, hrCuarto: Double?): String? {
        val generadas = motor().evaluarAtemperado(ContextoAtemperado(tempCuarto, hrCuarto))
        return generadas.takeIf { it.isNotEmpty() }
            ?.joinToString(" ") { "${it.quePaso}. ${it.queHacer}" }
    }

    // ------------------------------------------------------------ empaque

    suspend fun guardarEmpaque(
        loteProduccionId: String,
        barras: Int,
        pesoUnitarioG: Double,
        fechaElaboracion: Instant = Instant.now(),
        vidaUtilMeses: Int = 12,
        etiqueta: Map<String, String> = emptyMap(),
    ): String {
        val vencimiento = fechaElaboracion
            .atZone(ZoneId.systemDefault())
            .plusMonths(vidaUtilMeses.toLong())
            .toInstant()

        val e = EmpaqueEntidad(
            loteProduccionId = loteProduccionId,
            barras = barras,
            pesoUnitarioG = pesoUnitarioG,
            fechaElaboracion = fechaElaboracion,
            fechaVencimiento = vencimiento,
            vidaUtilMeses = vidaUtilMeses,
            etiquetaJson = json.encodeToString(etiqueta),
            codigoQr = porId(loteProduccionId)?.codigo.orEmpty(),
        )

        bd.withTransaction {
            dao.guardarEmpaque(e)
            porId(loteProduccionId)?.let { tanda ->
                dao.actualizar(
                    tanda.copy(
                        kgChocolate = barras * pesoUnitarioG / 1000.0,
                        estado = "terminado",
                        comunes = tanda.comunes.copy(
                            modificadoEn = Instant.now(),
                            estadoSync = EstadoSync.PENDIENTE,
                        ),
                    ),
                )
            }
            bd.inventario().insertarMovimiento(
                MovimientoInventarioEntidad(
                    item = "chocolate",
                    tipo = "entrada",
                    cantidad = barras.toDouble(),
                    unidad = "barras",
                    fecha = fechaElaboracion,
                    referencia = loteProduccionId,
                ),
            )
        }
        sync.encolar("empaques", e.id, "crear")
        return e.id
    }

    /** Todo el recorrido de una barra hasta su finca de origen (RF-LOT-04). */
    suspend fun trazabilidadDe(loteProduccionId: String): List<String> {
        val c = completa(loteProduccionId) ?: return emptyList()
        val lineas = mutableListOf("Tanda ${c.produccion.codigo}")

        for (o in c.origenes) {
            val lote = bd.lotes().porId(o.loteId) ?: continue
            val finca = lote.fincaId?.let { config.fincaPorId(it) }
            lineas += buildString {
                append("Lote ${lote.codigo} · ${o.kgUsados} kg")
                finca?.let { append(" · ${it.nombre}") }
            }
        }
        return lineas
    }
}

/** Una tanda con todo lo que cuelga de ella. */
data class ProduccionCompleta(
    val produccion: LoteProduccionEntidad,
    val origenes: List<LoteProduccionOrigenEntidad> = emptyList(),
    val tostado: TostadoEntidad? = null,
    val descascarillado: DescascarilladoEntidad? = null,
    val refinado: RefinadoEntidad? = null,
    val atemperados: List<AtemperadoEntidad> = emptyList(),
    val empaque: EmpaqueEntidad? = null,
) {
    val kgGranoUsado: Double get() = origenes.sumOf { it.kgUsados }
}

/** No se puede usar un lote con la venta bloqueada (RN-15). */
class LoteBloqueado(val loteId: String, val motivo: String) : IllegalStateException(
    "Ese lote está bloqueado y no se puede usar en producción: $motivo",
)

File: referencia/flutter-app/lib/datos/repositorios/repositorio_produccion.dart
/// Repositorio del lote de producción: del grano seco a la barra empacada.
library;

import 'dart:convert';

import 'package:drift/drift.dart';

import '../../nucleo/calculo/receta.dart';
import '../../nucleo/modelo/etapas.dart';
import '../../nucleo/reglas/alerta.dart';
import '../../nucleo/reglas/motor_reglas.dart';
import '../bd/base_datos.dart';
import '../bd/tablas/comunes.dart';
import 'repositorio_alertas.dart';
import 'repositorio_base.dart';
import 'repositorio_configuracion.dart';

/// Una tanda con todo lo que cuelga de ella.
class ProduccionCompleta {
  const ProduccionCompleta({
    required this.produccion,
    this.origenes = const [],
    this.tostado,
    this.descascarillado,
    this.refinado,
    this.atemperados = const [],
    this.empaque,
  });

  final LoteProduccion produccion;
  final List<LoteProduccionOrigen> origenes;
  final Tostado? tostado;
  final Descascarillado? descascarillado;
  final Refinado? refinado;
  final List<Atemperado> atemperados;
  final Empaque? empaque;

  double get kgGranoUsado =>
      origenes.fold<double>(0, (a, o) => a + o.kgUsados);
}

class RepositorioProduccion extends RepositorioBase {
  RepositorioProduccion(
    super.bd,
    super.sync, {
    required this.config,
    required this.alertas,
  });

  final RepositorioConfiguracion config;
  final RepositorioAlertas alertas;

  Future<MotorReglas> _motor() async => MotorReglas(await config.umbrales());

  Stream<List<LoteProduccion>> observar() => (bd.select(bd.lotesProduccion)
        ..where((t) => t.eliminado.equals(false))
        ..orderBy([
          (t) => OrderingTerm(expression: t.fecha, mode: OrderingMode.desc)
        ]))
      .watch();

  Future<LoteProduccion?> porId(String id) =>
      (bd.select(bd.lotesProduccion)..where((t) => t.id.equals(id)))
          .getSingleOrNull();

  /// Crea una tanda a partir de uno o varios lotes de grano (RF-LOT-05).
  ///
  /// Descuenta el grano del inventario, porque salir de almacén hacia
  /// producción es un movimiento real (RF-ALM-03).
  Future<LoteProduccion> crear({
    required Map<String, double> kgPorLote,
    double porcentajeCacao = 90,
  }) async {
    if (kgPorLote.isEmpty) {
      throw ArgumentError('Hay que indicar al menos un lote de grano');
    }
    for (final entrada in kgPorLote.entries) {
      final bloqueo = await alertas.bloqueoActivo(entrada.key);
      if (bloqueo != null) {
        throw LoteBloqueado(entrada.key, bloqueo.quePaso);
      }
    }

    final ahora = DateTime.now();
    final existentes = await bd.select(bd.lotesProduccion).get();
    final codigo = siguienteCodigo(
        'P', ahora.year, existentes.map((p) => p.codigo).toList());
    final id = uuidGenerador.v4();

    await bd.transaction(() async {
      await bd.into(bd.lotesProduccion).insert(
            LotesProduccionCompanion.insert(
              id: Value(id),
              codigo: codigo,
              fecha: ahora,
              porcentajeCacao: Value(porcentajeCacao),
            ),
          );
      for (final entrada in kgPorLote.entries) {
        await bd.into(bd.lotesProduccionOrigen).insert(
              LotesProduccionOrigenCompanion.insert(
                loteProduccionId: id,
                loteId: entrada.key,
                kgUsados: entrada.value,
              ),
            );
        await bd.into(bd.movimientosInventario).insert(
              MovimientosInventarioCompanion.insert(
                item: 'grano_seco',
                tipo: 'salida',
                cantidad: entrada.value,
                fecha: ahora,
                referencia: Value(codigo),
              ),
            );
      }
    });
    await encolar('lotes_produccion', id, 'crear', carga: {'codigo': codigo});
    return (await porId(id))!;
  }

  Future<ProduccionCompleta?> completa(String id) async {
    final p = await porId(id);
    if (p == null) return null;
    return ProduccionCompleta(
      produccion: p,
      origenes: await (bd.select(bd.lotesProduccionOrigen)
            ..where((t) => t.loteProduccionId.equals(id)))
          .get(),
      tostado: await (bd.select(bd.tostados)
            ..where((t) =>
                t.loteProduccionId.equals(id) & t.eliminado.equals(false)))
          .getSingleOrNull(),
      descascarillado: await (bd.select(bd.descascarillados)
            ..where((t) =>
                t.loteProduccionId.equals(id) & t.eliminado.equals(false)))
          .getSingleOrNull(),
      refinado: await (bd.select(bd.refinados)
            ..where((t) =>
                t.loteProduccionId.equals(id) & t.eliminado.equals(false)))
          .getSingleOrNull(),
      atemperados: await (bd.select(bd.atemperados)
            ..where((t) =>
                t.loteProduccionId.equals(id) & t.eliminado.equals(false))
            ..orderBy([(t) => OrderingTerm(expression: t.fecha)]))
          .get(),
      empaque: await (bd.select(bd.empaques)
            ..where((t) =>
                t.loteProduccionId.equals(id) & t.eliminado.equals(false)))
          .getSingleOrNull(),
    );
  }

  // ------------------------------------------------------------ tostado

  /// Registra el tostado y evalúa la merma (RN-12).
  Future<List<Alerta>> guardarTostado({
    required String loteProduccionId,
    String? equipoId,
    required double kgEntrada,
    double? kgSalida,
    double? tempC,
    int? minutos,
    List<Map<String, dynamic>> perfil = const [],
    String gradoIa = '',
    double? confianzaIa,
    String gradoUsuario = '',
    String modeloVersion = '',
    String? fotoId,
    String recetaNombre = '',
  }) async {
    final existente = await (bd.select(bd.tostados)
          ..where((t) => t.loteProduccionId.equals(loteProduccionId)))
        .getSingleOrNull();
    final id = existente?.id ?? uuidGenerador.v4();

    await bd.into(bd.tostados).insertOnConflictUpdate(
          TostadosCompanion.insert(
            id: Value(id),
            loteProduccionId: loteProduccionId,
            equipoId: Value(equipoId),
            fecha: DateTime.now(),
            kgEntrada: Value(kgEntrada),
            kgSalida: Value(kgSalida),
            tempC: Value(tempC),
            minutos: Value(minutos),
            perfilJson: Value(jsonEncode(perfil)),
            gradoIa: Value(gradoIa),
            confianzaIa: Value(confianzaIa),
            gradoUsuario: Value(gradoUsuario),
            modeloVersion: Value(modeloVersion),
            fotoId: Value(fotoId),
            recetaNombre: Value(recetaNombre),
            modificadoEn: Value(DateTime.now()),
          ),
        );
    await encolar('tostados', id, 'actualizar');

    if (kgSalida == null) return const [];
    final motor = await _motor();
    final generadas = motor.evaluarTostado(
        ContextoTostado(kgEntrada: kgEntrada, kgSalida: kgSalida));
    return alertas.registrar(generadas, loteProduccionId: loteProduccionId);
  }

  // ------------------------------------------------------- descascarillado

  Future<List<Alerta>> guardarDescascarillado({
    required String loteProduccionId,
    required double kgNibs,
    required double kgCascarilla,
    String notas = '',
  }) async {
    final existente = await (bd.select(bd.descascarillados)
          ..where((t) => t.loteProduccionId.equals(loteProduccionId)))
        .getSingleOrNull();
    final id = existente?.id ?? uuidGenerador.v4();

    await bd.into(bd.descascarillados).insertOnConflictUpdate(
          DescascarilladosCompanion.insert(
            id: Value(id),
            loteProduccionId: loteProduccionId,
            fecha: DateTime.now(),
            kgNibs: Value(kgNibs),
            kgCascarilla: Value(kgCascarilla),
            notas: Value(notas),
            modificadoEn: Value(DateTime.now()),
          ),
        );
    await encolar('descascarillados', id, 'actualizar');
    await bd.into(bd.movimientosInventario).insert(
          MovimientosInventarioCompanion.insert(
            item: 'nibs',
            tipo: 'entrada',
            cantidad: kgNibs,
            fecha: DateTime.now(),
            referencia: Value(loteProduccionId),
          ),
        );

    final tostado = await (bd.select(bd.tostados)
          ..where((t) => t.loteProduccionId.equals(loteProduccionId)))
        .getSingleOrNull();
    if (tostado == null || tostado.kgSalida == null) return const [];

    final motor = await _motor();
    final generadas = motor.evaluarTostado(ContextoTostado(
      kgEntrada: tostado.kgEntrada,
      kgSalida: tostado.kgSalida!,
      kgNibs: kgNibs,
      kgCascarilla: kgCascarilla,
    ));
    return alertas.registrar(generadas, loteProduccionId: loteProduccionId);
  }

  // ------------------------------------------------------------ refinado

  /// Calcula la receta para los nibs disponibles (RF-REF-01).
  Receta calcularReceta({
    required double kgNibs,
    double porcentajeCacao = 90,
    double mantecaExtraPct = 0,
    bool usarLecitina = true,
  }) =>
      const CalculadoraReceta().calcular(
        kgNibs: kgNibs,
        porcentajeCacao: porcentajeCacao,
        mantecaExtraPct: mantecaExtraPct,
        usarLecitina: usarLecitina,
      );

  Future<String> guardarRefinado({
    required String loteProduccionId,
    required DateTime inicio,
    DateTime? fin,
    double? horas,
    required double nibsKg,
    required double azucarKg,
    double mantecaKg = 0,
    double lecitinaKg = 0,
    DateTime? momentoAzucar,
    double? tempC,
    Map<String, int> sensorial = const {},
    String notas = '',
  }) async {
    final existente = await (bd.select(bd.refinados)
          ..where((t) => t.loteProduccionId.equals(loteProduccionId)))
        .getSingleOrNull();
    final id = existente?.id ?? uuidGenerador.v4();

    await bd.into(bd.refinados).insertOnConflictUpdate(
          RefinadosCompanion.insert(
            id: Value(id),
            loteProduccionId: loteProduccionId,
            inicio: inicio,
            fin: Value(fin),
            horas: Value(horas),
            nibsKg: Value(nibsKg),
            azucarKg: Value(azucarKg),
            mantecaKg: Value(mantecaKg),
            lecitinaKg: Value(lecitinaKg),
            momentoAzucar: Value(momentoAzucar),
            tempC: Value(tempC),
            sensorialJson: Value(jsonEncode(sensorial)),
            notas: Value(notas),
            modificadoEn: Value(DateTime.now()),
          ),
        );
    await encolar('refinados', id, 'actualizar');

    // El azúcar y la manteca salen del inventario al usarse.
    for (final par in {
      'azucar': azucarKg,
      'manteca': mantecaKg,
      'lecitina': lecitinaKg,
    }.entries) {
      if (par.value > 0) {
        await bd.into(bd.movimientosInventario).insert(
              MovimientosInventarioCompanion.insert(
                item: par.key,
                tipo: 'salida',
                cantidad: par.value,
                fecha: DateTime.now(),
                referencia: Value(loteProduccionId),
              ),
            );
      }
    }
    return id;
  }

  // ------------------------------------------------------------ atemperado

  /// Registra el atemperado y evalúa RN-13 y RN-14.
  Future<List<Alerta>> guardarAtemperado({
    required String loteProduccionId,
    MetodoAtemperado metodo = MetodoAtemperado.siembra,
    Map<String, double> temperaturas = const {},
    double? tempCuarto,
    double? hrCuarto,
    bool? pruebaPapel,
    String resultadoIa = '',
    double? confianzaIa,
    String resultadoUsuario = '',
    String modeloVersion = '',
    String? fotoId,
    int? diasInspeccion,
  }) async {
    final id = uuidGenerador.v4();
    await bd.into(bd.atemperados).insert(
          AtemperadosCompanion.insert(
            id: Value(id),
            loteProduccionId: loteProduccionId,
            fecha: DateTime.now(),
            metodo: Value(metodo),
            tempsJson: Value(jsonEncode(temperaturas)),
            tempCuarto: Value(tempCuarto),
            hrCuarto: Value(hrCuarto),
            pruebaPapel: Value(pruebaPapel),
            resultadoIa: Value(resultadoIa),
            confianzaIa: Value(confianzaIa),
            resultadoUsuario: Value(resultadoUsuario),
            modeloVersion: Value(modeloVersion),
            fotoId: Value(fotoId),
            diasInspeccion: Value(diasInspeccion),
          ),
        );
    await encolar('atemperados', id, 'crear');

    final motor = await _motor();
    final generadas = motor.evaluarAtemperado(ContextoAtemperado(
      tempCuartoC: tempCuarto,
      humedadCuartoPct: hrCuarto,
      tempTrabajoC: temperaturas['trabajo'],
    ));
    return alertas.registrar(generadas, loteProduccionId: loteProduccionId);
  }

  /// ¿Se puede moldear ahora mismo en estas condiciones? (RF-ATE-02)
  Future<String?> advertenciaDelCuarto(
      {double? tempCuarto, double? hrCuarto}) async {
    final motor = await _motor();
    final generadas = motor.evaluarAtemperado(ContextoAtemperado(
        tempCuartoC: tempCuarto, humedadCuartoPct: hrCuarto));
    if (generadas.isEmpty) return null;
    return generadas.map((a) => '${a.quePaso}. ${a.queHacer}').join(' ');
  }

  // ------------------------------------------------------------ empaque

  Future<String> guardarEmpaque({
    required String loteProduccionId,
    required int barras,
    required double pesoUnitarioG,
    DateTime? fechaElaboracion,
    int vidaUtilMeses = 12,
    Map<String, dynamic> etiqueta = const {},
  }) async {
    final elaboracion = fechaElaboracion ?? DateTime.now();
    final vencimiento = DateTime(
      elaboracion.year,
      elaboracion.month + vidaUtilMeses,
      elaboracion.day,
    );
    final id = uuidGenerador.v4();

    await bd.transaction(() async {
      await bd.into(bd.empaques).insert(
            EmpaquesCompanion.insert(
              id: Value(id),
              loteProduccionId: loteProduccionId,
              barras: Value(barras),
              pesoUnitarioG: Value(pesoUnitarioG),
              fechaElaboracion: elaboracion,
              fechaVencimiento: vencimiento,
              vidaUtilMeses: Value(vidaUtilMeses),
              etiquetaJson: Value(jsonEncode(etiqueta)),
              codigoQr: Value(loteProduccionId),
            ),
          );
      await (bd.update(bd.lotesProduccion)
            ..where((t) => t.id.equals(loteProduccionId)))
          .write(LotesProduccionCompanion(
        kgChocolate: Value(barras * pesoUnitarioG / 1000),
        estado: const Value('terminado'),
        modificadoEn: Value(DateTime.now()),
        estadoSync: const Value(EstadoSync.pendiente),
      ));
      await bd.into(bd.movimientosInventario).insert(
            MovimientosInventarioCompanion.insert(
              item: 'chocolate',
              tipo: 'entrada',
              cantidad: barras.toDouble(),
              unidad: const Value('barras'),
              fecha: elaboracion,
              referencia: Value(loteProduccionId),
            ),
          );
    });
    await encolar('empaques', id, 'crear');
    return id;
  }

  /// Todo el recorrido de una barra hasta su finca de origen (RF-LOT-04).
  Future<List<String>> trazabilidadDe(String loteProduccionId) async {
    final c = await completa(loteProduccionId);
    if (c == null) return const [];
    final lineas = <String>['Tanda ${c.produccion.codigo}'];

    for (final o in c.origenes) {
      final lote = await (bd.select(bd.lotes)..where((t) => t.id.equals(o.loteId)))
          .getSingleOrNull();
      if (lote == null) continue;
      final finca = lote.fincaId == null
          ? null
          : await (bd.select(bd.fincas)..where((t) => t.id.equals(lote.fincaId!)))
              .getSingleOrNull();
      lineas.add('Lote ${lote.codigo} · ${o.kgUsados} kg'
          '${finca == null ? '' : ' · ${finca.nombre}'}');
    }
    return lineas;
  }
}

/// No se puede usar un lote con la venta bloqueada (RN-15).
class LoteBloqueado implements Exception {
  const LoteBloqueado(this.loteId, this.motivo);
  final String loteId;
  final String motivo;

  @override
  String toString() =>
      'Ese lote está bloqueado y no se puede usar en producción: $motivo';
}
